design/one-pager-easy-stacks.md

```diff
@@ -0,0 +1,294 @@
+  something usually means touching a few different files. While this may not seem
+  like a big hurdle, changing some high level parameters of that environment is
+  cumbersome _after_ it's deployed.
+* A user who is familiar with AWS may not be familiar with Azure or GCP; so,
+  ready-made configuration stacks that promise the exact same setup would be
+  useful for them to see how they can achieve similar setups and run various
+  benchmarks according to their business needs.
+* After you run `kubectl apply -f <directory of YAMLs>`, a dependency graph of
+  resources starts to resolve. However, it's not always clear for user to know _when_
+  everything is actually ready to use unless they know readiness of which resources
+  signal the readiness of the whole environment.
+
+## Goals
+
+The main goal of this design is to make it easier for people to write sets of
+configurations and be able to manage those configuration sets on a higher level.
+
+It is important that the design puts forward:
+
+* A base boilerplate package where creating a new configuration set that has a
+  controller is as easy as changing the YAML files.
+* A way to declare pre-requisites that needs to be satisfied by the cluster
+  for the configuration set to work.
+* A way to change a high level configuration by changing 1 value that is then
+  translated to the YAMLs that are already deployed.
+
+## Proposal
+
+Crossplane has the notion of stacks that can be used as extension to Crossplane's
+abilites, to deploy an application and all other things that you can do with
+Kubernetes controllers since main components of stacks are CRDs and controllers
+watching them.
+
+To achieve the mentioned goals, we can publish the YAML files in a stack with a
+controller that has its own CRD with high level parameters. Controller deploys
+and updates the YAMLs that are packaged in the stack image according to user's
+input on the CR instance of that stack that user creates.
+
+As sky is the limit when it comes to stacks, we need to put down some general
+rules to be adhered by initial configuration stacks.
+
+* CRDs of the resources that will be deployed by the stack are declared as
+  dependency so that user knows what CR instances will be deployed by the given
+  configuration stack.
+* It should be possible to have different sets of configurations in one stack
+  and user should be able to choose from them via the CR that is an instance
+  of stack's CRD.
+  * Note that it's highly preferred to have 1 custom resource definition that
+    the stack uses and expose the high level variations of the parameters through
+    `spec` instead of having different custom resource definitions for each class
+    of configuration.
+* A minimal high level set of configurations are exposed to the user through
+  the stack's CR, such as `Provider` reference to refer in the deployed resource
+  or `region`.
+* Controller should signal readiness of all resources in CR's status.
```

Comment:
What is `readiness`? How does the controller know they are ready?

Reply:
Do we get much value from modelling the fact that the entire configuration set is ready?

Resource classes have no concept of readiness - if they exist they're ready - so I imagine this would be communicating that the various supporting managed resources (VPCs, etc) were ready to use. Will anything block on this readiness (e.g. you can't create a `KubernetesApplication` until the `EasyAWS` is ready), or would it just be a signal to the infrastructure operator that everything came up okay?

Reply:
> What is readiness? How does the controller know they are ready?

Its details are in `Reconciliation` section.

> Do we get much value from modelling the fact that the entire configuration set is ready?

Thinking about it again, not a lot really. It'd help the infra-owner to know everything is ready but that'd be just informational. I'll remove that part.

Reply:
Oh, actually I think if we remove that there will be no `status`, meaning no way for anyone to know the stuff is actually working. So, I think we'd want to have at least the `conditions` part of the status with reconciliation status.

Reply:
It does seem valuable to have a way to check the status of everything in a single place. However, I am not sure we need to make it a requirement for resource packs. Since the main goal is the focus on applying a set of configuration best practices without writing your own, that goal seems somewhat orthogonal to the goal of seeing the statuses in one place.

Put another way, I think it is an improvement even without the status aggregation, so we don't necessarily need to include the status aggregation in the first version. For example, it is possible that we add status/readiness aggregation in some other way, with a UI or with some `trace`-like functionality.

Reply:
@suskin I have added `ConditionedStatus` to status of the initial gcp implementation. We'll have `Synced` type condition to be set by the generic controller.